Lib/test/pickletester.py

```diff
@@ -2076,15 +2095,39 @@ def test_framing_many_objects(self):
 
     def test_framing_large_objects(self):
         N = 1024 * 1024
-        obj = [b'x' * N, b'y' * N, b'z' * N]
+        obj = [b'x' * N, b'y' * N, 'z' * N]
         for proto in range(4, pickle.HIGHEST_PROTOCOL + 1):
-            with self.subTest(proto=proto):
-                pickled = self.dumps(obj, proto)
-                unpickled = self.loads(pickled)
-                self.assertEqual(obj, unpickled)
-                n_frames = count_opcode(pickle.FRAME, pickled)
-                self.assertGreaterEqual(n_frames, len(obj))
-                self.check_frame_opcodes(pickled)
+            for fast in [True, False]:
+                if fast and not hasattr(self, 'pickler'):
+                    continue
+
+                # The default picklers do not include large binary objects in
+                # frames.
+                # However some alternative implementation of dump / dumps can
+                # output pickles that include large binary objects inside
+                # protocol 4 frames. The test classes of such implementations
+                # make it explicit by setting the following flag.
+                frameless_blobs = getattr(self, 'frameless_blobs', True)
+
+                with self.subTest(proto=proto, fast=fast):
+                    if hasattr(self, 'pickler'):
+                        buf = io.BytesIO()
+                        pickler = self.pickler(buf, protocol=proto)
+                        pickler.fast = fast
+                        pickler.dump(obj)
+                        pickled = buf.getvalue()
+                    else:
+                        pickled = self.dumps(obj, proto)
+                    unpickled = self.loads(pickled)
+                    self.assertEqual(obj, unpickled)
+                    n_frames = count_opcode(pickle.FRAME, pickled)
+                    if not fast:
+                        # One frame per memoize for each large object.
```

Comment:
I suggest to not write a frame for a data <= 3 bytes. Reading a frame requires 3 reading operations, and 3 bytes can be read by at most 3 operations. There is no benefit of using a frame.

Reply:
I'm going to create a separate issue for more accurate optimization of frames, but maybe the simpler heuristic (size <= 3) can be used in this PR. If this will not complicate the code.

Reply:
I'd rather see the new heuristics in a separate PR personally :-)

Reply:
I also rather do it in a separate PR as it's not trivial to remove a frame header from the `output_buffer` a posteriori.